src/operator/nn/mkldnn/mkldnn_rnn_impl.h

```diff
@@ -437,91 +450,118 @@ static void MKLDNNRNNForwardUnidi(bool state_outputs,
-        AdjustGruWeightGateOrder(wh, H, H);
-        AdjustGruBiasGateOrder(b_ptr, H);
-        AdjustGruBiasGateOrder(b_ptr + H * ngates, H);
+        AdjustGruWeightGateOrder(wx, input_size, hidden_size);
+        AdjustGruWeightGateOrder(wh, hidden_size, hidden_size);
+        has_adjusted = true;
       }
       src_wx_f.set_data_handle(wx);
       src_wh_f.set_data_handle(wh);
     } else {
-      for (int l = 0; l < L; l++) {
-        DType* wx = w_ptr;
-        DType* wh = w_ptr + I * H * ngates;
-        DType* bx = b_ptr + l * ngates * H * 2;
-        DType* bh = b_ptr + l * ngates * H * 2 + H * ngates;
+      for (int l = 0; l < num_layer; l++) {
+        DType* wx = w_ptr + l * w_size;
+        DType* wh = wx + input_size * hidden_size * ngates;
         if (mode == rnn_enum::kGru) {
-          AdjustGruWeightGateOrder(wx, I, H);
-          AdjustGruWeightGateOrder(wh, H, H);
-          AdjustGruBiasGateOrder(bx, H);
-          AdjustGruBiasGateOrder(bh, H);
+          AdjustGruWeightGateOrder(wx, input_size, hidden_size);
+          AdjustGruWeightGateOrder(wh, hidden_size, hidden_size);
+          has_adjusted = true;
         }
         srcs_data_x.push_back(wx);
         srcs_data_h.push_back(wh);
         src_l_dim_x.push_back(weights_layer_r_tz);
         src_l_dim_h.push_back(weights_iter_r_tz);
-        w_ptr = w_ptr + w_size;
       }
       ConcatData(mkldnn::memory::format::ldgoi, mkldnn::memory::format::ldgoi,
-          src_l_dim_x, weights_layer_tz, mkldnn_dtype, 0, srcs_data_x, src_wx_f);
+          src_l_dim_x, weights_layer_tz, mkldnn_dtype, 0, srcs_data_x, src_wx_f,
+          &(mkldnn_mems->weight_layer_mems));
       ConcatData(mkldnn::memory::format::ldgoi, mkldnn::memory::format::ldgoi,
-          src_l_dim_h, weights_iter_tz, mkldnn_dtype, 0, srcs_data_h, src_wh_f);
+          src_l_dim_h, weights_iter_tz, mkldnn_dtype, 0, srcs_data_h, src_wh_f,
+          &(mkldnn_mems->weight_iter_mems));
     }
-    MKLDNNStream::Get()->RegisterPrim(reorder(src_wx_f, (*wx_memory)[layer_index]));
-    MKLDNNStream::Get()->RegisterPrim(reorder(src_wh_f, (*wh_memory)[layer_index]));
-
-    DType* user_bias_f = reinterpret_cast<DType *> ((*bias_memory)[layer_index].get_data_handle());
-    #pragma omp parallel for num_threads(omp_threads)
-    for (int j = 0; j < L * single_b_size; j++) {
-      int k = j / single_b_size;
-      user_bias_f[j] = b_ptr[j + k * single_b_size] + b_ptr[j + k * single_b_size + single_b_size];
+    MKLDNNStream::Get()->RegisterPrim(reorder(src_wx_f, mkldnn_mems->wx_memory[layer_index]));
+    MKLDNNStream::Get()->RegisterPrim(reorder(src_wh_f, mkldnn_mems->wh_memory[layer_index]));
+
+    DType* user_bias_f = reinterpret_cast<DType *>(
+        mkldnn_mems->bias_memory[layer_index].get_data_handle());
+    if (mode == rnn_enum::kGru) {
+      const int mx_single_b_sz = ngates * hidden_size;
+      for (int l = 0; l < num_layer; l++) {
+        #pragma omp parallel for num_threads(omp_threads)
```

Comment:
We could use collapse(2) for these two for loops instead of only the inner loop. But note that Microsoft Visual C++ compiler might not support `collapse`, which could be separated by Macro.

Reply:
Thanks for noting that.